dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -80,40 +80,52 @@ static LimitedChannel pinUntilError(List<LimitedChannel> channels) {
      * @deprecated prefer {@link #pinUntilError}
      */
     @Deprecated
-    static LimitedChannel pinUntilErrorWithoutReshuffle(List<LimitedChannel> channels) {
+    static LimitedChannel pinUntilErrorWithoutReshuffle(List<CompositeLimitedChannel> channels) {
         ConstantNodeList constantNodeList = new ConstantNodeList(channels, SafeThreadLocalRandom.get());
         return new PinUntilErrorChannel(constantNodeList);
     }
 
     @Override
     public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
         int currentIndex = currentHost.get();
-        LimitedChannel channel = nodeList.get(currentIndex);
+        CompositeLimitedChannel channel = nodeList.get(currentIndex);
 
-        Optional<ListenableFuture<Response>> maybeFuture = channel.maybeExecute(endpoint, request);
-        if (!maybeFuture.isPresent()) {
-            OptionalInt next = incrementHostIfNecessary(currentIndex);
-            debugLogCurrentChannelRejected(currentIndex, channel, next);
-            return Optional.empty(); // if the caller retries immediately, we'll get the next host
-        }
+        LimitedResponse limitedResponse = channel.maybeExecute(endpoint, request);
+        return limitedResponse.matches(new LimitedResponse.Cases<Optional<ListenableFuture<Response>>>() {
+            @Override
+            public Optional<ListenableFuture<Response>> blacklisted() {
+                OptionalInt next = incrementHostIfNecessary(currentIndex);
+                debugLogCurrentChannelRejected(currentIndex, channel, next);
+                return maybeExecute(endpoint, request);
+            }
 
-        ListenableFuture<Response> future = maybeFuture.get();
-        return Optional.of(DialogueFutures.addDirectCallback(future, new FutureCallback<Response>() {
             @Override
-            public void onSuccess(Response response) {
-                if (Responses.isQosStatus(response) || Responses.isServerError(response)) {
-                    OptionalInt next = incrementHostIfNecessary(currentIndex);
-                    debugLogReceivedErrorStatus(currentIndex, channel, response, next);
-                    // TODO(dfox): handle 308 See Other somehow, as we currently don't have a host -> channel mapping
-                }
+            public Optional<ListenableFuture<Response>> limited() {
+                debugLogLimitedRequest(currentIndex, channel);
+                return Optional.empty();
```

Comment:
what does it semantically mean for the PinUntilErrorChannel to return Optional.empty? If an inner concurrency limiter decided this host was overloaded, why are we not selecting another host here?

Reply:
I think this also explains why https://github.com/palantir/dialogue/blob/fo/differentiate-limits-blacklist/simulation/src/test/resources/report.md#drastic_slowdownconcurrency_limiter_pin_until_error doesn't look good after the slow node was reverted.